test/js/slice.spec.js

```diff
@@ -0,0 +1,133 @@
+/**
+ * Unless explicitly stated otherwise all files in this repository are licensed under the Apache-2.0 License.
+ * This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2022 Datadog, Inc.
+ **/
+const { TaintedUtils, taintFormattedString, formatTaintedValue } = require('./util')
+const assert = require('assert')
+
+describe('Slice', function () {
+  const id = '1'
+
+  const testCases = [
+    {
+      source: ':+-foobarbaz-+:',
+      result: ':+-foo-+:',
+      start: 0,
+      end: 3
+    },
+    {
+      source: ':+-foobarbaz-+:',
+      result: ':+-foo-+:',
+      start: -9,
+      end: 3
+    },
+    {
+      source: ':+-foobarbaz-+:',
+      result: ':+-bar-+:',
+      start: 3,
+      end: 6
+    },
+    {
+      source: ':+-foobarbaz-+:',
+      result: ':+-baz-+:',
+      start: 6,
+      end: 9
+    },
+    {
+      source: ':+-foobarbaz-+:',
+      result: ':+-baz-+:',
+      start: 6
+    },
+    {
+      source: 'foo:+-bar-+:baz',
+      result: 'foo',
+      start: 0,
+      end: 3
+    },
+    {
+      source: 'foo:+-bar-+:baz',
+      result: 'oo:+-ba-+:',
+      start: 1,
+      end: 5
+    },
+    {
+      source: 'foo:+-bar-+:baz',
+      result: ':+-r-+:baz',
+      start: 5,
+      end: 9
+    },
+    {
+      source: 'foo:+-bar-+:baz',
+      result: 'baz',
+      start: 6,
+      end: 9
+    },
+    {
+      source: 'foo:+-bar-+:b:+-az-+:',
+      result: 'b:+-az-+:',
+      start: 6,
+      end: 9
+    },
+    {
+      source: 'foo:+-bar-+:b:+-a-+:z',
+      result: 'b:+-a-+:z',
+      start: 6,
+      end: 9
+    },
+    {
+      source: 'foo:+-bar-+:b:+-a-+:z:+-z-+::+-z-+:',
+      result: 'b:+-a-+:z',
+      start: 6,
+      end: 9
```

Comment:
I miss a test where `end` is bigger than `source.length`

Reply:
And also, I mess some test with more than 1 range in the result, something like `:+-r-+:b:+-a-+:z:+-z-+:` in the result, for example

Reply:
And we need other test, to test that 1 char length result is new instance and we are not tainting a literal.

Reply:
Added more tests